src/splits.cpp

```diff
@@ -637,9 +638,58 @@ RawMatrix pack_splits_logical_vec(LogicalVector x) {
     
     unsigned char byte = 0;
     for (int k = 0; k < limit; ++k) {
-      if (p[k] == TRUE) byte |= (1u << k);
     }
     out(0, b) = byte;
   }
   return out;
 }
+
+// Fast count of unique splits in a phylogenetic tree
+// Optimized replacement for collapse.singles(x)[["Nnode"]] - 1L - TreeIsRooted(x)
+// [[Rcpp::export]]
+int cpp_count_splits(const Rcpp::IntegerMatrix& edge, int nTip) {
+  if (edge.ncol() != 2) {
+    Rcpp::stop("Edge matrix must contain two columns");
+  }
+  
+  const int n_edge = edge.nrow();
+  if (n_edge == 0 || nTip < 4) {
+    return 0;
+  }
+  
+  // Get all unique parent nodes - these represent internal nodes
+  std::set<int> internal_nodes;
+  
+  for (int i = 0; i < n_edge; ++i) {
+    const int parent = edge(i, 0);
+    const int child = edge(i, 1);
+    
+    // Parent must be an internal node (> nTip)
+    if (parent > nTip) {
+      internal_nodes.insert(parent);
+    }
+  }
+  
+  const int n_internal = internal_nodes.size();
+  
+  // Check if tree is rooted: count children of root node
+  // Root is the minimum parent node
```

Comment:
Usually, but not necessarily: This assumes that a tree is sensibly numbered, which is guaranteed for tree in "order = 'preorder'" but not for an arbitrary tree